src/main/java/org/spongepowered/api/command/parameter/CommandContext.java

```diff
@@ -0,0 +1,168 @@
+     *
+     * @return {@code true} if so
+     */
+    boolean isCompletion();
+
+    /**
+     * Returns whether this context has any value for the given argument key.
+     *
+     * @param key The key to look up
+     * @return whether there are any values present
+     */
+    boolean hasAny(Text key);
+
+    /**
+     * Returns whether this context has any value for the given argument key.
+     *
+     * @param key The key to look up
+     * @return whether there are any values present
+     */
+    boolean hasAny(String key);
+
+    /**
+     * Gets the value for the given key if the key has only one value.
+     *
+     * @param key the key to get
+     * @param <T> the expected type of the argument
+     * @return the argument
+     */
+    <T> Optional<T> getOne(Text key);
+
+    /**
+     * Gets the value for the given key if the key has only one value.
+     *
+     * @param key the key to get
+     * @param <T> the expected type of the argument
+     * @return the argument
+     */
+    <T> Optional<T> getOne(String key);
+
+    /**
+     * Gets the value for the given key if the key has only one value,
+     * otherwise, throws a {@link NoSuchElementException}.
+     *
+     * @param key the key to get
+     * @param <T> the expected type of the argument
+     * @return the argument
+     */
+    <T> T getOneUnchecked(Text key);
+
+    /**
+     * Gets the value for the given key if the key has only one value,
+     * otherwise, throws a {@link NoSuchElementException}.
+     *
+     * @param key the key to get
+     * @param <T> the expected type of the argument
+     * @return the argument
+     */
+    <T> T getOneUnchecked(String key);
+
+    /**
+     * Gets all values for the given argument. May return an empty list if no
+     * values are present.
+     *
+     * @param key The key to get values for
+     * @param <T> the type of value to get
+     * @return the collection of all values
+     */
+    <T> Collection<T> getAll(Text key);
+
+    /**
+     * Gets all values for the given argument. May return an empty list if no
+     * values are present.
+     *
+     * @param key The key to get values for
+     * @param <T> the type of value to get
+     * @return the collection of all values
+     */
+    <T> Collection<T> getAll(String key);
+
+    /**
+     * Adds a parsed object into the context, for use by commands.
+     *
+     * @param key The key to store the entry under.
+     * @param object The object to store.
+     */
+    void putEntry(Text key, Object object);
+
+    /**
+     * Adds a parsed object into the context, for use by commands.
+     *
+     * @param key The key to store the entry under.
+     * @param object The object to store.
+     */
+    void putEntry(String key, Object object);
+
+    /**
+     * Creates a snapshot of this context and returns a state object that can
+     * be used to restore the context to this state.
+     *
+     * @return The state.
+     */
+    Snapshot getState();
```

Comment:
I _might_ call the snapshot a `<blah>State` instead, because snapshots have been used in the API as inspectable... well, snapshots. There should be no need to inspect these - it’s possibly woth making that clear down to the naming.